// @aws-sdk/client-kafka-node/commands/CreateConfigurationCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var __aws_sdk_middleware_stack = require("@aws-sdk/middleware-stack");
var CreateConfiguration_1 = require("../model/operations/CreateConfiguration");
var CreateConfigurationCommand = /** @class */ (function () {
    function CreateConfigurationCommand(input) {
        this.input = input;
        this.model = CreateConfiguration_1.CreateConfiguration;
        this.middlewareStack = new __aws_sdk_middleware_stack.MiddlewareStack();
    }
    CreateConfigurationCommand.prototype.resolveMiddleware = function (clientStack, configuration) {
        var handler = configuration.handler;
        var stack = clientStack.concat(this.middlewareStack);
        var handlerExecutionContext = {
            logger: {},
            model: this.model
        };
        return stack.resolve(handler(handlerExecutionContext), handlerExecutionContext);
    };
    return CreateConfigurationCommand;
}());
exports.CreateConfigurationCommand = CreateConfigurationCommand;
