// @aws-sdk/client-kafka-node/commands/DescribeConfigurationCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var __aws_sdk_middleware_stack = require("@aws-sdk/middleware-stack");
var DescribeConfiguration_1 = require("../model/operations/DescribeConfiguration");
var DescribeConfigurationCommand = /** @class */ (function () {
    function DescribeConfigurationCommand(input) {
        this.input = input;
        this.model = DescribeConfiguration_1.DescribeConfiguration;
        this.middlewareStack = new __aws_sdk_middleware_stack.MiddlewareStack();
    }
    DescribeConfigurationCommand.prototype.resolveMiddleware = function (clientStack, configuration) {
        var handler = configuration.handler;
        var stack = clientStack.concat(this.middlewareStack);
        var handlerExecutionContext = {
            logger: {},
            model: this.model
        };
        return stack.resolve(handler(handlerExecutionContext), handlerExecutionContext);
    };
    return DescribeConfigurationCommand;
}());
exports.DescribeConfigurationCommand = DescribeConfigurationCommand;
