// @aws-sdk/client-kafka-node/commands/ListTagsForResourceCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var __aws_sdk_middleware_stack = require("@aws-sdk/middleware-stack");
var ListTagsForResource_1 = require("../model/operations/ListTagsForResource");
var ListTagsForResourceCommand = /** @class */ (function () {
    function ListTagsForResourceCommand(input) {
        this.input = input;
        this.model = ListTagsForResource_1.ListTagsForResource;
        this.middlewareStack = new __aws_sdk_middleware_stack.MiddlewareStack();
    }
    ListTagsForResourceCommand.prototype.resolveMiddleware = function (clientStack, configuration) {
        var handler = configuration.handler;
        var stack = clientStack.concat(this.middlewareStack);
        var handlerExecutionContext = {
            logger: {},
            model: this.model
        };
        return stack.resolve(handler(handlerExecutionContext), handlerExecutionContext);
    };
    return ListTagsForResourceCommand;
}());
exports.ListTagsForResourceCommand = ListTagsForResourceCommand;
