// @aws-sdk/client-kafka-node/commands/ListConfigurationRevisionsCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var __aws_sdk_middleware_stack = require("@aws-sdk/middleware-stack");
var ListConfigurationRevisions_1 = require("../model/operations/ListConfigurationRevisions");
var ListConfigurationRevisionsCommand = /** @class */ (function () {
    function ListConfigurationRevisionsCommand(input) {
        this.input = input;
        this.model = ListConfigurationRevisions_1.ListConfigurationRevisions;
        this.middlewareStack = new __aws_sdk_middleware_stack.MiddlewareStack();
    }
    ListConfigurationRevisionsCommand.prototype.resolveMiddleware = function (clientStack, configuration) {
        var handler = configuration.handler;
        var stack = clientStack.concat(this.middlewareStack);
        var handlerExecutionContext = {
            logger: {},
            model: this.model
        };
        return stack.resolve(handler(handlerExecutionContext), handlerExecutionContext);
    };
    return ListConfigurationRevisionsCommand;
}());
exports.ListConfigurationRevisionsCommand = ListConfigurationRevisionsCommand;
