// @aws-sdk/client-kafka-node/commands/ListConfigurationsCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var __aws_sdk_middleware_stack = require("@aws-sdk/middleware-stack");
var ListConfigurations_1 = require("../model/operations/ListConfigurations");
var ListConfigurationsCommand = /** @class */ (function () {
    function ListConfigurationsCommand(input) {
        this.input = input;
        this.model = ListConfigurations_1.ListConfigurations;
        this.middlewareStack = new __aws_sdk_middleware_stack.MiddlewareStack();
    }
    ListConfigurationsCommand.prototype.resolveMiddleware = function (clientStack, configuration) {
        var handler = configuration.handler;
        var stack = clientStack.concat(this.middlewareStack);
        var handlerExecutionContext = {
            logger: {},
            model: this.model
        };
        return stack.resolve(handler(handlerExecutionContext), handlerExecutionContext);
    };
    return ListConfigurationsCommand;
}());
exports.ListConfigurationsCommand = ListConfigurationsCommand;
