// @aws-sdk/client-kafka-node/commands/ListClusterOperationsCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var __aws_sdk_middleware_stack = require("@aws-sdk/middleware-stack");
var ListClusterOperations_1 = require("../model/operations/ListClusterOperations");
var ListClusterOperationsCommand = /** @class */ (function () {
    function ListClusterOperationsCommand(input) {
        this.input = input;
        this.model = ListClusterOperations_1.ListClusterOperations;
        this.middlewareStack = new __aws_sdk_middleware_stack.MiddlewareStack();
    }
    ListClusterOperationsCommand.prototype.resolveMiddleware = function (clientStack, configuration) {
        var handler = configuration.handler;
        var stack = clientStack.concat(this.middlewareStack);
        var handlerExecutionContext = {
            logger: {},
            model: this.model
        };
        return stack.resolve(handler(handlerExecutionContext), handlerExecutionContext);
    };
    return ListClusterOperationsCommand;
}());
exports.ListClusterOperationsCommand = ListClusterOperationsCommand;
