// @aws-sdk/client-kafka-node/commands/UpdateClusterConfigurationCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var __aws_sdk_middleware_stack = require("@aws-sdk/middleware-stack");
var UpdateClusterConfiguration_1 = require("../model/operations/UpdateClusterConfiguration");
var UpdateClusterConfigurationCommand = /** @class */ (function () {
    function UpdateClusterConfigurationCommand(input) {
        this.input = input;
        this.model = UpdateClusterConfiguration_1.UpdateClusterConfiguration;
        this.middlewareStack = new __aws_sdk_middleware_stack.MiddlewareStack();
    }
    UpdateClusterConfigurationCommand.prototype.resolveMiddleware = function (clientStack, configuration) {
        var handler = configuration.handler;
        var stack = clientStack.concat(this.middlewareStack);
        var handlerExecutionContext = {
            logger: {},
            model: this.model
        };
        return stack.resolve(handler(handlerExecutionContext), handlerExecutionContext);
    };
    return UpdateClusterConfigurationCommand;
}());
exports.UpdateClusterConfigurationCommand = UpdateClusterConfigurationCommand;
