// @aws-sdk/client-kafka-node/model/operations/CreateConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var CreateConfigurationInput_1 = require("../shapes/CreateConfigurationInput");
var CreateConfigurationOutput_1 = require("../shapes/CreateConfigurationOutput");
var BadRequestException_1 = require("../shapes/BadRequestException");
var InternalServerErrorException_1 = require("../shapes/InternalServerErrorException");
var UnauthorizedException_1 = require("../shapes/UnauthorizedException");
var ForbiddenException_1 = require("../shapes/ForbiddenException");
var ServiceUnavailableException_1 = require("../shapes/ServiceUnavailableException");
var TooManyRequestsException_1 = require("../shapes/TooManyRequestsException");
var ConflictException_1 = require("../shapes/ConflictException");
var ServiceMetadata_1 = require("../ServiceMetadata");
exports.CreateConfiguration = {
    metadata: ServiceMetadata_1.ServiceMetadata,
    name: "CreateConfiguration",
    http: {
        method: "POST",
        requestUri: "/v1/configurations"
    },
    input: {
        shape: CreateConfigurationInput_1.CreateConfigurationInput
    },
    output: {
        shape: CreateConfigurationOutput_1.CreateConfigurationOutput
    },
    errors: [
        {
            shape: BadRequestException_1.BadRequestException
        },
        {
            shape: InternalServerErrorException_1.InternalServerErrorException
        },
        {
            shape: UnauthorizedException_1.UnauthorizedException
        },
        {
            shape: ForbiddenException_1.ForbiddenException
        },
        {
            shape: ServiceUnavailableException_1.ServiceUnavailableException
        },
        {
            shape: TooManyRequestsException_1.TooManyRequestsException
        },
        {
            shape: ConflictException_1.ConflictException
        }
    ]
};
