// @aws-sdk/client-kafka-node/model/operations/DescribeConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DescribeConfigurationInput_1 = require("../shapes/DescribeConfigurationInput");
var DescribeConfigurationOutput_1 = require("../shapes/DescribeConfigurationOutput");
var BadRequestException_1 = require("../shapes/BadRequestException");
var UnauthorizedException_1 = require("../shapes/UnauthorizedException");
var InternalServerErrorException_1 = require("../shapes/InternalServerErrorException");
var ForbiddenException_1 = require("../shapes/ForbiddenException");
var NotFoundException_1 = require("../shapes/NotFoundException");
var ServiceUnavailableException_1 = require("../shapes/ServiceUnavailableException");
var ServiceMetadata_1 = require("../ServiceMetadata");
exports.DescribeConfiguration = {
    metadata: ServiceMetadata_1.ServiceMetadata,
    name: "DescribeConfiguration",
    http: {
        method: "GET",
        requestUri: "/v1/configurations/{arn}"
    },
    input: {
        shape: DescribeConfigurationInput_1.DescribeConfigurationInput
    },
    output: {
        shape: DescribeConfigurationOutput_1.DescribeConfigurationOutput
    },
    errors: [
        {
            shape: BadRequestException_1.BadRequestException
        },
        {
            shape: UnauthorizedException_1.UnauthorizedException
        },
        {
            shape: InternalServerErrorException_1.InternalServerErrorException
        },
        {
            shape: ForbiddenException_1.ForbiddenException
        },
        {
            shape: NotFoundException_1.NotFoundException
        },
        {
            shape: ServiceUnavailableException_1.ServiceUnavailableException
        }
    ]
};
