// @aws-sdk/client-kafka-node/model/operations/GetBootstrapBrokers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var GetBootstrapBrokersInput_1 = require("../shapes/GetBootstrapBrokersInput");
var GetBootstrapBrokersOutput_1 = require("../shapes/GetBootstrapBrokersOutput");
var BadRequestException_1 = require("../shapes/BadRequestException");
var UnauthorizedException_1 = require("../shapes/UnauthorizedException");
var InternalServerErrorException_1 = require("../shapes/InternalServerErrorException");
var ConflictException_1 = require("../shapes/ConflictException");
var ForbiddenException_1 = require("../shapes/ForbiddenException");
var ServiceMetadata_1 = require("../ServiceMetadata");
exports.GetBootstrapBrokers = {
    metadata: ServiceMetadata_1.ServiceMetadata,
    name: "GetBootstrapBrokers",
    http: {
        method: "GET",
        requestUri: "/v1/clusters/{clusterArn}/bootstrap-brokers"
    },
    input: {
        shape: GetBootstrapBrokersInput_1.GetBootstrapBrokersInput
    },
    output: {
        shape: GetBootstrapBrokersOutput_1.GetBootstrapBrokersOutput
    },
    errors: [
        {
            shape: BadRequestException_1.BadRequestException
        },
        {
            shape: UnauthorizedException_1.UnauthorizedException
        },
        {
            shape: InternalServerErrorException_1.InternalServerErrorException
        },
        {
            shape: ConflictException_1.ConflictException
        },
        {
            shape: ForbiddenException_1.ForbiddenException
        }
    ]
};
