// @aws-sdk/client-kafka-node/model/operations/UpdateBrokerStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var UpdateBrokerStorageInput_1 = require("../shapes/UpdateBrokerStorageInput");
var UpdateBrokerStorageOutput_1 = require("../shapes/UpdateBrokerStorageOutput");
var ServiceUnavailableException_1 = require("../shapes/ServiceUnavailableException");
var BadRequestException_1 = require("../shapes/BadRequestException");
var UnauthorizedException_1 = require("../shapes/UnauthorizedException");
var InternalServerErrorException_1 = require("../shapes/InternalServerErrorException");
var ForbiddenException_1 = require("../shapes/ForbiddenException");
var ServiceMetadata_1 = require("../ServiceMetadata");
exports.UpdateBrokerStorage = {
    metadata: ServiceMetadata_1.ServiceMetadata,
    name: "UpdateBrokerStorage",
    http: {
        method: "PUT",
        requestUri: "/v1/clusters/{clusterArn}/nodes/storage"
    },
    input: {
        shape: UpdateBrokerStorageInput_1.UpdateBrokerStorageInput
    },
    output: {
        shape: UpdateBrokerStorageOutput_1.UpdateBrokerStorageOutput
    },
    errors: [
        {
            shape: ServiceUnavailableException_1.ServiceUnavailableException
        },
        {
            shape: BadRequestException_1.BadRequestException
        },
        {
            shape: UnauthorizedException_1.UnauthorizedException
        },
        {
            shape: InternalServerErrorException_1.InternalServerErrorException
        },
        {
            shape: ForbiddenException_1.ForbiddenException
        }
    ]
};
