// @aws-sdk/client-kafka-node/model/shapes/_BrokerNodeInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _BrokerSoftwareInfo_1 = require("./_BrokerSoftwareInfo");
var ___listOf__string_1 = require("./___listOf__string");
exports._BrokerNodeInfo = {
    type: "structure",
    required: [],
    members: {
        AttachedENIId: {
            shape: {
                type: "string"
            },
            locationName: "attachedENIId"
        },
        BrokerId: {
            shape: {
                type: "float"
            },
            locationName: "brokerId"
        },
        ClientSubnet: {
            shape: {
                type: "string"
            },
            locationName: "clientSubnet"
        },
        ClientVpcIpAddress: {
            shape: {
                type: "string"
            },
            locationName: "clientVpcIpAddress"
        },
        CurrentBrokerSoftwareInfo: {
            shape: _BrokerSoftwareInfo_1._BrokerSoftwareInfo,
            locationName: "currentBrokerSoftwareInfo"
        },
        Endpoints: {
            shape: ___listOf__string_1.___listOf__string,
            locationName: "endpoints"
        }
    }
};
