// @aws-sdk/client-kafka-node/model/shapes/_BrokerNodeGroupInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ___listOf__string_1 = require("./___listOf__string");
var _StorageInfo_1 = require("./_StorageInfo");
exports._BrokerNodeGroupInfo = {
    type: "structure",
    required: ["ClientSubnets", "InstanceType"],
    members: {
        BrokerAZDistribution: {
            shape: {
                type: "string"
            },
            locationName: "brokerAZDistribution"
        },
        ClientSubnets: {
            shape: ___listOf__string_1.___listOf__string,
            locationName: "clientSubnets"
        },
        InstanceType: {
            shape: {
                type: "string",
                min: 5
            },
            locationName: "instanceType"
        },
        SecurityGroups: {
            shape: ___listOf__string_1.___listOf__string,
            locationName: "securityGroups"
        },
        StorageInfo: {
            shape: _StorageInfo_1._StorageInfo,
            locationName: "storageInfo"
        }
    }
};
