// @aws-sdk/client-kafka-node/model/shapes/_Configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ___listOf__string_1 = require("./___listOf__string");
var _ConfigurationRevision_1 = require("./_ConfigurationRevision");
exports._Configuration = {
    type: "structure",
    required: [
        "Description",
        "LatestRevision",
        "CreationTime",
        "KafkaVersions",
        "Arn",
        "Name"
    ],
    members: {
        Arn: {
            shape: {
                type: "string"
            },
            locationName: "arn"
        },
        CreationTime: {
            shape: {
                type: "timestamp",
                timestampFormat: "iso8601"
            },
            locationName: "creationTime"
        },
        Description: {
            shape: {
                type: "string"
            },
            locationName: "description"
        },
        KafkaVersions: {
            shape: ___listOf__string_1.___listOf__string,
            locationName: "kafkaVersions"
        },
        LatestRevision: {
            shape: _ConfigurationRevision_1._ConfigurationRevision,
            locationName: "latestRevision"
        },
        Name: {
            shape: {
                type: "string"
            },
            locationName: "name"
        }
    }
};
