// @aws-sdk/client-kafka-node/model/shapes/_NodeInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _BrokerNodeInfo_1 = require("./_BrokerNodeInfo");
var _ZookeeperNodeInfo_1 = require("./_ZookeeperNodeInfo");
exports._NodeInfo = {
    type: "structure",
    required: [],
    members: {
        AddedToClusterTime: {
            shape: {
                type: "string"
            },
            locationName: "addedToClusterTime"
        },
        BrokerNodeInfo: {
            shape: _BrokerNodeInfo_1._BrokerNodeInfo,
            locationName: "brokerNodeInfo"
        },
        InstanceType: {
            shape: {
                type: "string"
            },
            locationName: "instanceType"
        },
        NodeARN: {
            shape: {
                type: "string"
            },
            locationName: "nodeARN"
        },
        NodeType: {
            shape: {
                type: "string"
            },
            locationName: "nodeType"
        },
        ZookeeperNodeInfo: {
            shape: _ZookeeperNodeInfo_1._ZookeeperNodeInfo,
            locationName: "zookeeperNodeInfo"
        }
    }
};
