// @aws-sdk/client-kafka-node/model/shapes/CreateClusterInput.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _BrokerNodeGroupInfo_1 = require("./_BrokerNodeGroupInfo");
var _ClientAuthentication_1 = require("./_ClientAuthentication");
var _ConfigurationInfo_1 = require("./_ConfigurationInfo");
var _EncryptionInfo_1 = require("./_EncryptionInfo");
var ___mapOf__string_1 = require("./___mapOf__string");
exports.CreateClusterInput = {
    type: "structure",
    required: [
        "BrokerNodeGroupInfo",
        "KafkaVersion",
        "NumberOfBrokerNodes",
        "ClusterName"
    ],
    members: {
        BrokerNodeGroupInfo: {
            shape: _BrokerNodeGroupInfo_1._BrokerNodeGroupInfo,
            locationName: "brokerNodeGroupInfo"
        },
        ClientAuthentication: {
            shape: _ClientAuthentication_1._ClientAuthentication,
            locationName: "clientAuthentication"
        },
        ClusterName: {
            shape: {
                type: "string",
                min: 1
            },
            locationName: "clusterName"
        },
        ConfigurationInfo: {
            shape: _ConfigurationInfo_1._ConfigurationInfo,
            locationName: "configurationInfo"
        },
        EncryptionInfo: {
            shape: _EncryptionInfo_1._EncryptionInfo,
            locationName: "encryptionInfo"
        },
        EnhancedMonitoring: {
            shape: {
                type: "string"
            },
            locationName: "enhancedMonitoring"
        },
        KafkaVersion: {
            shape: {
                type: "string",
                min: 1
            },
            locationName: "kafkaVersion"
        },
        NumberOfBrokerNodes: {
            shape: {
                type: "integer",
                min: 1
            },
            locationName: "numberOfBrokerNodes"
        },
        Tags: {
            shape: ___mapOf__string_1.___mapOf__string,
            locationName: "tags"
        }
    }
};
